### ISSUE TITLE
Custom repository doesn't register

### ISSUE THREAD

**Issue by aerni** [statamic/cms#2126]
## Bug Description
This is a follow up on [the discussion](https://discordapp.com/channels/489818810157891584/674287924547813416/735871428116611112) with @jasonvarga on Discord.

I'm trying to extend the validation rules with a custom repository. The custom `EntryRepository` never gets hit. @damcclean mentioned, that it might be an issue with the namespace; that `Statamic` overrides `Aerni`. Don't know if that's the issue. But I thought I mention it.

**Custom EntryRepository:**
```php
<?php

namespace Aerni\Snipcart\Repositories;

use Statamic\Stache\Repositories\EntryRepository as StatamicEntryRepository;

class EntryRepository extends StatamicEntryRepository
{
    public function createRules($collection, $site)
    {
        $rules = parent::createRules($collection, $site);

        if ($collection->handle() == 'products') {
            $rules['sku'] = 'required|unique_entry_value:'.$collection->handle().',null,'.$site->handle();
        }

        return $rules;
    }

    public function updateRules($collection, $entry)
    {
        $rules = parent::updateRules($collection, $entry);

        if ($collection->handle() == 'products') {
            $rules['sku'] = 'required|unique_entry_value:'.$collection->handle().','.$entry->id().','.$entry->locale();
        }

        return $rules;
    }
}

```

**Service Provider:**

```php
public function register(): void
    {
        parent::register();
        $this->app->bind(Repositories\EntryRepository::class);
    }
```
## Environment

**Statamic version:** 3.0.0.-beta-39

**PHP version:** 7.4.6

**Install method:** Upgrade from beta-38

**Comment by jasonvarga** [comment]
`bind` accepts two arguments: the interface and then your class. You are only passing one.

```php
$this->app->bind(
    \Statamic\Contracts\Entries\EntryRepository::class,
    \Aerni\Snipcart\RepositoriesDatabaseEntryRepository::class
);
```

Although when I recommended this to you I didn't realize it was for an addon. I don't think changing the entry repository is a decision for an addon to make, it should be for the project. We can do better. However, it should work for now.

**Comment by aerni** [comment]
True, my bad! I tested it. It works when adding it to the `App\Providers\AppServiceProvider`. But in an addon Service Provider it doesn't work.

**Comment by jasonvarga** [comment]
Wrap it in Statamic::booted() so it happens after Statamic is done. Maybe Statamic was re-binding its one over the top of yours.

```php
Statamic::booted(function () {
    $this->app->bind(
        \Statamic\Contracts\Entries\EntryRepository::class,
        \Aerni\Snipcart\RepositoriesDatabaseEntryRepository::class
    );
});
```

**Comment by aerni** [comment]
I tried that already but it didn't work. I fiddled around with it a bit more and found that some other function in my Service Provider was somehow affecting the binding not to happen. I now switched the function order and it works as expected!

On that note: When does the `Statamic::booted()` event fire? After the `register()` or the `boot()` method? Like is it good practice to wrap functions in the `boot()` method that depend on Statamic into `Statamic::booted()`? Or do I only need this in the `register()` method?
